cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/FileBatchSource.java

```diff
@@ -0,0 +1,211 @@
+  private static final String TABLE_DESCRIPTION = "Name of the Table that keeps track of the last time files " +
+    "were read in.";
+  private static final String INPUT_FORMAT_CLASS_DESCRIPTION = "Name of the input format class, which must be a " +
+    "subclass of FileInputFormat. Defaults to TextInputFormat";
+  private static final String FILESYSTEM_DESCRIPTION = "Distributed file system to read in from.";
+  private static final Gson GSON = new Gson();
+  private static final Logger LOG = LoggerFactory.getLogger(FileBatchSource.class);
+  private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd-HH-mm");
+  private static final Schema DEFAULT_SCHEMA = Schema.recordOf(
+    "event",
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("body", Schema.of(Schema.Type.STRING))
+  );
+
+  private final FileBatchConfig config;
+  private KeyValueTable table;
+  private Date prevMinute;
+
+  public FileBatchSource(FileBatchConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    if (config.timeTable != null) {
+      pipelineConfigurer.createDataset(config.timeTable, KeyValueTable.class, DatasetProperties.EMPTY);
+    }
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    //calculate date one minute ago, rounded down to the nearest minute
+    prevMinute = new Date(context.getLogicalStartTime() - TimeUnit.MINUTES.toMillis(1));
+    Calendar cal = Calendar.getInstance();
+    cal.setTime(prevMinute);
+    cal.set(Calendar.SECOND, 0);
+    cal.set(Calendar.MILLISECOND, 0);
+    prevMinute = cal.getTime();
+
+    Job job = context.getHadoopJob();
+    if (config.fileSystemProperties != null) {
+      FileBatchProperties properties = GSON.fromJson(config.fileSystemProperties, FileBatchProperties.class);
+      for (Map.Entry<String, FileBatchProperties.FileProperty> entry : properties.getPropertyMap().entrySet()) {
+        FileBatchProperties.FileProperty property = entry.getValue();
+        job.getConfiguration().set(property.getPropertyName(), property.getPropertyValue());
+      }
+    }
+
+    if (config.fileRegex != null) {
+      job.getConfiguration().set(INPUT_REGEX_CONFIG, config.fileRegex);
+    }
+    job.getConfiguration().set(INPUT_NAME_CONFIG, config.path);
+
+    if (config.timeTable != null) {
+      table = context.getDataset(config.timeTable);
+      String lastTimeRead = Bytes.toString(table.read("lastTimeRead"));
+      if (lastTimeRead == null) {
+        lastTimeRead = "0";
+      }
+      job.getConfiguration().set(LAST_TIME_READ, lastTimeRead);
+    }
+
+    job.getConfiguration().set(CUTOFF_READ_TIME, DATE_FORMAT.format(prevMinute));
+    if (config.inputFormatClass != null) {
+      ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+      Class<?> classType = classLoader.loadClass(config.inputFormatClass);
```

Comment:
I think you missed my previous comment here. Make this type ``Class<? extends FileInputFormat>``, so you don't need lines 142-147 below.